### ISSUE TITLE
Tracked objects besides gamepads

### ISSUE THREAD

**Issue by dmarcos** [immersive-web/webxr#133]
Valve is opening their lighthouse tracking technology and we will see soon a lot of tracked objects besides controllers. How are WebVR APIs going to accommodate those?

**Comment by toji** [comment]
I doubt that we can reasonably create a spec that intelligently handles every possible type of tracked object. I'd rather see them logically grouped by use (Displays, Controllers, etc.) One very simple step we can take, though, is define Pose in such a way that it can trivially be used by any object with tracking capabilities. (In contrast to the current situation where WebVR and Gamepad have their own concept of Pose)

**Comment by dmarcos** [comment]
What group would for instance a chair or my tracked pet fall in? How will we query all the tracked objects available? Do we need a separate API for this?

**Comment by blairmacintyre** [comment]
It seems like there are two separate issues here:
- finding out what things can be tracked
- how they report back

In the VR research world, this is not solved.  Libraries like VRPN try to at least provide a generic interface to "sensors";  it seems like it would be nice if there was a general "web 6DOF sensor API"  that was provided a uniform way to at least retrieve and work with the data.  The specific kinds of tracking systems (be they connected to the machine/device like Lighthouse would be, or externally running, like we might have with a room-based tracker, or implemented inside the application via a library, such as a computer vision tracker might be) could then be set up and configured in whatever way they need to be, but could report in in a compatible way.

In argon.js/Argon4, we massage anything that can report a 6DOF frame of reference into a Cesium Entity (for example), and relate them to each other in one big hierarchy.

This does not seem like a thing WebVR should solve;  it seems like it might be an external standard/extension for sensors, and WebVR would be able to be configured to _use_ one of those sensors as it's head tracker.

**Comment by kearwood** [comment]
With the recent announcement of the HTC tracker device, we can expect to soon see an infinite variety of tracked devices that also have a variable number of axis, buttons, and other sensors.  The number of tracked devices at once will also increase to "dozens" rather than just a couple at once.

The HTC tracker specifically doesn't establish relationships in a hierarchy as VRPN does; however, this functionality may eventually be useful as devices plugged into the trackers (via the included USB port or pogo pins) may spawn a tree of sensors. (One prototype demonstrated at CES was a glove with tracked fingers attached to an HTC tracker).

How would everyone feel about enumerating all of these tracked devices as "Gamepads" with a variable number of buttons and axis, while their VRPN-like  relationships be exposed later with a parallel structure referencing the Gamepad objects at each node?  This would allow us to proceed on our current path while not precluding VRPN-like association later..

**Comment by anssiko** [comment]
@blairmacintyre wrote:

>Libraries like VRPN try to at least provide a generic interface to "sensors"; it seems like it would be nice if there was a general "web 6DOF sensor API" that was provided a uniform way to at least retrieve and work with the data.

>This does not seem like a thing WebVR should solve; it seems like it might be an external standard/extension for sensors, and WebVR would be able to be configured to use one of those sensors as it's head tracker.

The [Generic Sensor API][1]  is solving exactly that. It provides a generic interface to sensors and a uniform way to retrieve and work with the sensor data.

Timing of this discussion is perfect. The `Sensor` base class is stabilizing, and we're shifting our focus to `MotionSensor` class that specifically aims to address WebVR use cases. Initially, we plan to add to `MotionSensor` the alternative representations (quaternion, rotation matrix), a generic BYOB solution, and means to handle batching.

That said, we'd like to understand WebVR use cases and requirements better to make sure we'll do the right thing considering reuse in WebVR, so please chime in with your requirements.

If you want to get a feel of the Generic Sensor-based APIs in a browser, flip the chrome://flags/#enable-generic-sensor flag in Chrome to try the Generic Sensor-based motion sensors: [Accelerometer][2], [Gyroscope][3], [Magnetometer][4]. Note these APIs will evolve as we add more features such as those described above.

Tagging @tobie and @pozdnyakov working with me on the specs & implementation.

[1]: https://w3c.github.io/sensors/
[2]: https://w3c.github.io/accelerometer/
[3]: https://w3c.github.io/gyroscope/
[4]: https://w3c.github.io/magnetometer/

**Comment by machenmusik** [comment]
> How would everyone feel about enumerating all of these tracked devices as "Gamepads" 

One problem here is that some browsers have finite limits on the number of simultaneous gamepads they expose; for example, Chromium only exposes an array of four, and actually there are problems when one has built-in devices that show up as gamepads and then tries to expose the full complement of Oculus input devices (remote, left touch, right touch, xbox gamepad).  If that limitation can be removed or expanded then this might be more viable as a short-term workaround.

But that masks the further problem of understanding what the tracked objects actually are, and should be doing...

At the moment we only have "OpenVR Gamepad" (or briefly "OpenVR Controller"); it is not immediately clear whether corresponding changes in OpenVR/SteamVR would be needed to improve that situation.  Applications may need methods of selecting which tracked object with Pose fulfills a desired role, similar to selecting which camera for getUserMedia().

**Comment by cvan** [comment]
See https://crbug.com/644028 for the bug w/ Chromium limiting to only four Gamepads.

**Comment by blairmacintyre** [comment]
> That said, we'd like to understand WebVR use cases and requirements better to make sure we'll do the right thing considering reuse in WebVR, so please chime in with your requirements.

The implementors can probably chime in more, but I'd expect the biggest issues will be synchronization (having high precision time stamps use a common time) and possibly some way to get all the reports in a queue (not just the most recent ones), so that they can be fed into a motion-model if need be (e.g., for neck modeling, for example).  

For the AR use case (where we might have video), it would be great, eventually if we had a way of time stamping video frames in a synchronized way, so we know when in the sensor stream a video frame was acquired.  I suspect this is beyond the scope of just the sensor api, though.

**Comment by tobie** [comment]
> The implementors can probably chime in more, but I'd expect the biggest issues will be synchronization (having high precision time stamps use a common time)

In the generic sensor API, this is so far handled by converting the timestamp provided by the underlying platform into monotonic time using `DOMHighResTimeStamp`. Comments on whether that's appropriate and fits your use cases would be welcomed.

> and possibly some way to get all the reports in a queue (not just the most recent ones),

We're in the process of designing a BYOB-based buffering API, so now's just the right time to chime in. Current plan is to update the buffer right before `requestAnimationFrame` (with more than one sample at a time for polling frequencies > 60 Hz). Would that meet your use cases? 

> so that they can be fed into a motion-model if need be (e.g., for neck modeling, for example).

How many samples would you need in such a case? What formats would you want those samples to be in? I'd love to see pseudo code of the kind of algorithms you'd be running (e.g. neck modelling).

> For the AR use case (where we might have video), it would be great, eventually if we had a way of time stamping video frames in a synchronized way, so we know when in the sensor stream a video frame was acquired. I suspect this is beyond the scope of just the sensor api, though.

I would imagine `DOMHighResTimeStamp` is sufficiently precisely defined to make that possible. But is it?

**Comment by anssiko** [comment]
Timestamp (represented as `DOMHighResTimeStamp`) was stripped out of `VRFrameData` recently (#175) and "can be added back in later after further discussion to more precisely define what the timestamp is relative to."

So it seems [`DOMHighResTimeStamp`][1] was not a perfect fit for the WebVR API as is.

(We're now tracking WebVR timestamp requirements to Generic Sensor API in https://github.com/w3c/sensors/issues/155.)

[1]: http://w3c.github.io/hr-time/#dom-domhighrestimestamp

**Comment by tobie** [comment]
@toji: just reading through #175. Looked for minutes of the Kirkland F2F, but couldn't find them. Can you clarify whether you removed the timestamp because of intrinsic characteristics of `DOMHighResTimeStamp` (if so, which ones?) or because it wasn't clear whether the timestamp pertained to the current frame—something along the lines of `requestAnimationFrame`'s _now_ timestamp (see [§ 8.9 Animation Frames](https://html.spec.whatwg.org/multipage/webappapis.html#animation-frames) in HTML)—or to the time at which the sensor readings were acquired?

**Comment by pozdnyakov** [comment]
`DOMHighResTimeStamp` could be a good way to synchronize VR frames not only with each other but also with many other things like rAF callbacks arguments, `Performance.now()` calls and even WebAudio output. This however requires using the same [time origin](http://w3c.github.io/hr-time/#time-origin) and the formerly present `VRFrameData.timestamp` used its own time origin ("The timestamp
-starts at 0 the first time {{getFrameData()}} is invoked") which looks inconvenient.

**Comment by blairmacintyre** [comment]
I really hope that we can keep consistent, meaningful timestamps in every part of the system.  Without them, we can't synchronize across libraries.  For example, it will definitely be the case that people will want to integrate different kind of sensor data, perhaps from distributed machines, into VR (and AR) systems.  We really need to have a way of understanding the relationships between local sensors (VR HMDs and controllers, additional sensors, video frames and the resulting tracking results, etc) as well as (eventually) data _between_ machines.  I understand that using `DOMHighResTimeStamp` doesn't solve the later problem, but certainly solving the local problem is at least a first step.

(I've been burned in the past, trying to integrate multiple libraries into one AR system, where the libraries either used their own definition of time, or didn't exposing accurate timing at all, so I'd like to vote for avoiding that problem here.)

**Comment by tobie** [comment]
> I understand that using DOMHighResTimeStamp doesn't solve the later problem, but certainly solving the local problem is at least a first step.

Well, combined with `performance.timing.navigationStart`, `DOMHighResTimeStamp` does let you get a Unix timestamp. This doesn't solve clock sync issues obviously, but it's a step forward.

**Comment by blairmacintyre** [comment]
I wondered about that, but I read a disclaimer about how `DOMHighResTimeStamp` increments based on real-world-time (e.g., at a constant rate) while the system time (exposed through things like `Date`) can be adjusted by the system for various reasons (NTP shifts, for example).  So, I could imagine a reliable distributed sync method would need to handle that.  It wouldn't be absurdly hard, of course, as distributed systems already deal with it, but I wasn't sure what other issues might be lurking.

As you say, though, using `DOMHighResTimeStamp` is a step in the right direction.

**Comment by tobie** [comment]
Afaik, you can calculate [clock drift](https://w3c.github.io/hr-time/#clock-drift) using something along the lines of:

```js
function drift() { return performance.timing.navigationStart + performance.now() - Date.now(); }
```

In theory, this should be enough to let us write a function that converts a `performance.timing.navigationStart`-based timestamp to an epoch-based one and vice versa.

In practice, it would be good to hear from the Web Perf folks on the topic. Pinging @igrigorik and @plehegar.

**Comment by tobie** [comment]
Sorry for hi-jacking this thread, btw. Would you want to split this conversation up somehow in a way that made more sense?

**Comment by machenmusik** [comment]
> But that masks the further problem of understanding what the tracked objects actually are, and should be doing...

Related to the looming issue, Nightly recently introduced Oculus Touch support, but with a different number of buttons and axes than Chromium reports for the same physical controllers.  that can only get worse with as other tracked objects are introduced.

Might be good to set forth some best practices guidelines for mappings e.g. button with analog value, purely digital button, clickable 2D pad/stick, 1D slider / throttle, 3DoF / 6DoF nodes (e.g. like a pair of nunchucks where each part is fully tracked, but they are one controller)...

**Comment by toji** [comment]
@tobie: There wasn't a problem with `DOMHighResTimeStamp` per say, so much as no consensus on what exactly the timestamp represented. The time that the pose was captured? If so, should it be different for the gamepads vs. the headset, because they're often polled at slightly different times. Or do you want all devices for a "frame" to have the same timestamp. (Sounds reasonable) If so what's generating that stamp? Must it match the timestamps generated by `requestAnimationFrame`? Keep in mind that those are often generated in a different place.

The concern was that there was a lot of unanswered questions and we didn't want people relying on the current haphazard behavior until they were resolved. Hence the timestamps were stripped out until we could decide exactly what they should represent. I do think that once that's settled thought we'll be using `DOMHighResTimeStamp`, though. I haven't seen a reason not to.

**Comment by blairmacintyre** [comment]
Thanks @toji.  I would tend to agree that unless they are well defined, in a useful time representation, they aren't useful.

The "right" answer (to me) is that "every device should have it's own timestamp that represents (as close as possible) what time the data represents".  So, the gamepads should have timestamps matching the time their state was last sampled (as should sensors that might be reported by the sensor API, or video frames, or device orientation, or geoposition, or ...).  If a driver is doing prediction and returning data that's been extrapolated forward, the timestamp should be the time that the data represents.  And so on.

I can't remember off the top of my head how the underlying VR APIs work, but if the "main" data provided by the underlying api that is returned by `getFrameData()` has a timestamp (i.e., we know what it was sampled), why wouldn't we just want to translate that to the same time scale as `DOMHighResTimeStamp` and return it?  Since the programmer can call `performance.now()` to get the _current_ time (in the same time representation), passing that in is not particularly useful.

**Comment by kearwood** [comment]
FWIW, the Firefox implementation of `VRFrameData.timeStamp` randomizes the initial timestamp value to some degree.  The delta of this value between frames is variable, corresponding to the variability of the values returned by the underlying VR API.

The randomization of the initial value on page load ensures that any details (ie. duration of VR session) are not leaked to web content.  The initial value starts low, so there is no risk of wrap-around and precision issues caused by large values.  It also encourages developers to make no assumptions about the initial value.

We still have freedom to define explicitly what the timeStamp values mean, as long as the deltas between frames are consistent with the underlying VR API deltas.

Perhaps we could sample `DOMHighResTimeStamp` at page load and synchronize the VRFrameData.timeStamp with an offset at that point.  If the gamepad API did the same, then there would be a single timeline to keep things simple.

There would still be a potential difference between the gamepad's timestamp and the `VRFrameData.timeStamp`...  The VRFrameData represents a predicted pose and may even represent a time in the future.  The Gamepad's pose may be sampled multiple times per frame by a drawing application, and could represent the instantaneous time of the sampled data.

We would need to ask some high-level questions before defining correlation between the timelines from DOMHighResTimestamp, VRFrameData.timeStamp and gamepad timestamp's...

- Can we define the VRFrameData.timeStamp to be the predicted time of presentation?  This may be in the future
- Can we define gamepad timestamps to match the sample time of the sensors?  This value may change on every call, potentially multiple times per frame.
- Will there need to be special consideration for underlying VR API's that do pose prediction on controllers?

**Comment by kearwood** [comment]
TL;DR - I think it's worthwhile unifying the timelines of the DOMHighResTimeStamp, WebVR, and Gamepad API's.  We just need to lock down the definition of the timestamps better before we can describe what their origin is

**Comment by machenmusik** [comment]
It seems like the obvious requirements for unified timestamping are
(1) monotonic non-decreasing
(2) sufficient precision to handle required frame / update rates
(3) convertible to/from real world wall clock time

At the moment, milliseconds seem sufficiently precise for (2).

Randomization to avoid leaking usage data is a nice touch but may cause ill effects unless relevant timestamps are indeed unified...

**Comment by tobie** [comment]
> Randomization to avoid leaking usage data is a nice touch but may cause ill effects unless relevant timestamps are indeed unified...

Yes, that has pretty serious implications when you integrate sensor values (iirc it creates drift).

**Comment by toji** [comment]
Hooo boy! This is an old issue, in'it? Come with me, dear commentators of the past, on a journey to discover the progress made since this issue was filed:

- Renamed the API to WebXR
- Overhauled how we track things, including introducing the ability to track arbitrary "XRSpaces".
- Revamped how input works entirely.
- Separated Gamepad input into an [entirely different module](https://immersive-web.github.io/webxr-gamepads-module/).
- Are just about to ship the API! Crazy, huh?

In practical terms, what this means for tracking of non-gamepad-like objects is this: If the thing you are tracking could reasonably be considered an input method (as is "used to make selections and initiate actions by the user") then it should be exposed as an `XRInputSource`. This is a generic interface that _can_ have gamepad data associated with it but is not required to do so, and also represents things like hand and gaze input. The input sources also have a `profiles` attribute that can be used to identify what type of input is being exposed.

For tracked objects that don't fit into the "input" bucket (such as anchors, markers, maybe skeletal tracking, etc.) the appropriate design pattern is to define a new interface that exposes an `XRSpace` that represents the tracked object with whatever additional data is appropriate. The `XRSpace` is the primitive that is passed into `XRFrame.getPose()` for tracking, and is a lot more flexible than what we previously had in WebVR.

Given that I think the above quite thoroughly addresses the original concerns of this issue, consider this issue fixed. :)
